File: server/app/schemas/todos.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel
from app.models.todo import TodoStatus

class TodoBase(BaseModel):
    title: str
    status: TodoStatus = TodoStatus.PENDING
    is_bookmarked: bool = False
    order: int = 0
    parent_id: Optional[int] = None

class TodoCreate(TodoBase):
    pass

class TodoUpdate(BaseModel):
    title: Optional[str] = None
    status: Optional[TodoStatus] = None
    is_bookmarked: Optional[bool] = None
    order: Optional[int] = None
    parent_id: Optional[int] = None

class TodoResponse(TodoBase):
    id: int
    user_id: int
    created_at: datetime
    modified_at: datetime
    subtasks: List['TodoResponse'] = []

    class Config:
        from_attributes = True

# Update forward references
TodoResponse.model_rebuild()

class TodoFilter(BaseModel):
    status: Optional[TodoStatus] = None
    is_bookmarked: Optional[bool] = None
    search: Optional[str] = None
    parent_id: Optional[int] = None
    columns: Optional[str] = None,

class PaginationParams(BaseModel):
    page: int = 1
    page_size: int = 10
    order_by: str = "created_at"
    order_direction: str = "desc" 

File: client/src/types/todos.ts
export type TodoStatus = 'pending' | 'in_progress' | 'completed';

export interface Todo {
  id: number;
  title: string;
  status: TodoStatus;
  is_bookmarked: boolean;
  order: number;
  parent_id: number | null;
  user_id: number;
  created_at: string;
  updated_at: string;
  subtasks: Todo[];
}

export interface TodoCreate {
  title: string;
  status: TodoStatus;
  is_bookmarked?: boolean;
  parent_id?: number | null;
}

export interface TodoUpdate {
  title?: string;
  status?: TodoStatus;
  is_bookmarked?: boolean;
  parent_id?: number | null;
}

export interface TodoFilter {
  status?: TodoStatus;
  is_bookmarked?: 'true' | 'false';
  parent_id?: number | null;
  search?: string;
}

export interface PaginationParams {
  page: number;
  page_size: number;
  order_by?: string;
  order_direction?: 'asc' | 'desc';
}

export interface TodoReorder {
  todo_id: number;
  new_order: number;
}

export interface TodoReorderRequest {
  reorders: TodoReorder[];
  parent_id?: number | null;
}

export interface PaginatedResponse<T> {
  items: T[];
  total_count: number;
  page: number;
  page_size: number;
} 